pkg/handlers/apk.go

```diff
@@ -0,0 +1,430 @@
+func readFile(file *zip.File) (io.ReadCloser, error) {
+	rc, err := file.Open()
+	if err != nil {
+		return nil, err
+	}
+	return rc, nil
+}
+
+// extractStringsFromResTable extracts the strings from the resources table
+// Note: This is a hacky way to get the strings from the resources table
+// APK strings are typically (always?) stored in the 0x7f000000-0x7fffffff range
+// https://chromium.googlesource.com/chromium/src/+/master/build/android/docs/life_of_a_resource.md
+func extractStringsFromResTable(resTable *apkparser.ResourceTable) (io.Reader, error) {
+	var resourceStrings bytes.Buffer
+	inStrings := false
+	for i := 0x7f000000; i <= 0x7fffffff; i++ {
+		entry, _ := resTable.GetResourceEntry(uint32(i))
+		if entry == nil {
+			continue
+		}
+		if entry.ResourceType == "string" {
+			inStrings = true
+			val, err := entry.GetValue().String()
+			if err != nil {
+				return nil, err
+			}
+			// Write directly to the buffer
+			_, err = resourceStrings.WriteString(fmt.Sprintf("%s: %s\n", entry.Key, val))
+			if err != nil {
+				return nil, err
+			}
+		}
+		// Exit the loop if we've finished processing the strings
+		if inStrings && entry.ResourceType != "string" {
+			break
+		}
+	}
+	return &resourceStrings, nil
+}
+
+// processDexFile decodes the dex file and returns the relevant instructions
+func processDexFile(ctx logContext.Context, rdr io.ReadCloser) (io.Reader, error) {
+	// dextk.Read() requires an io.ReaderAt interface,
+	// so we first convert the reader to a byte slice
+	data, err := io.ReadAll(rdr)
+	if err != nil {
+		return nil, err
+	}
+	bytesRdr := bytes.NewReader(data)
+
+	// Read the dex file
+	dexReader, err := dextk.Read(bytesRdr)
+	if err != nil {
+		return nil, err
+	}
+
+	defaultKeywords := DefaultDetectorKeywords()
+
+	// Get relevant instruction data from the dex file
+	var dexOutput bytes.Buffer
+	ci := dexReader.ClassIter()
+	for ci.HasNext() {
+		node, err := ci.Next()
+		if err != nil {
+			break
+		}
+		processDexClass(ctx, dexReader, node, defaultKeywords, &dexOutput)
+	}
+
+	return &dexOutput, nil
+}
+
+// processDexClass processes a single class node's methods
+func processDexClass(ctx logContext.Context, dexReader *dextk.Reader, node dextk.ClassNode, defaultKeywords map[string]struct{}, dexOutput *bytes.Buffer) {
+
+	var classOutput bytes.Buffer
+	methodValues := make(map[string]struct{})
+
+	// Process Direct Methods
+	processDexMethod(ctx, dexReader, node.DirectMethods, &classOutput, methodValues)
+	// Process Virtual Methods
+	processDexMethod(ctx, dexReader, node.VirtualMethods, &classOutput, methodValues)
+
+	// Write the classOutput to the dexOutput
+	dexOutput.Write(classOutput.Bytes())
+
+	// Stringify the classOutput value for case-insensitive keyword matching
+	classOutputLower := strings.ToLower(classOutput.String())
+
+	// Check if classOutput contains any of the default keywords
+	foundKeywords := make(map[string]struct{})
+	for keyword := range defaultKeywords {
```

Comment:
suggestion: This might be easier to explain synchronously, but during testing, I noticed this approach incurs a significant performance penalty. We’re iterating over all the keywords and running `strings.Contains` on each, which is costly.  

Here are a few thoughts:  
1. Your idea to move `DefaultDetectors` to the `detectors` package is a solid first step.  
2. We can adopt a more efficient data structure. Using the Aho-Corasick algorithm (similar to how we handle detectors) would allow us to leverage a trie for a substantial performance improvement—potentially by several orders of magnitude.  

This will require some changes to the handler, but a straightforward approach could look like this:  
```
// singleton instance and once initializer
var instance *KeywordMatcher

func init() {
	allDetectors := detectors.DefaultDetectors()
	// Remove keywords that cause lots of false positives.
	var exclusions = []string{
		"AKIA", "SG.", "pat", "token", "gh", "github", "sql", "database", "http", "key", "api-", "sdk-", "float", "-us", "gh", "pat", "token", "sid", "http", "private", "key", "segment", "close", "protocols", "verifier", "box", "privacy", "dm", "sl.", "vf", "flat",
	}

	var keywords []string
	exclusionSet := make(map[string]struct{})
	for _, excl := range exclusions {
		exclusionSet[strings.ToLower(excl)] = struct{}{}
	}

	// Aggregate all keywords from detectors.
	for _, detector := range allDetectors {
		for _, kw := range detector.Keywords() {
			kwLower := strings.ToLower(kw)
			if _, excluded := exclusionSet[kwLower]; !excluded {
				keywords = append(keywords, kwLower)
			}
		}
	}

	instance = &KeywordMatcher{trie: ahocorasick.NewTrieBuilder().AddStrings(keywords).Build()}
}

// KeywordMatcher encapsulates the Aho-Corasick trie for keyword matching.
type KeywordMatcher struct {
	mu   sync.RWMutex
	trie *ahocorasick.Trie
}

func getMatcher() *KeywordMatcher { return instance }

// FindKeywords scans the input text and returns a slice of matched keywords.
func (km *KeywordMatcher) FindKeywords(text []byte) []string {
	km.mu.RLock()
	defer km.mu.RUnlock()

	matches := km.trie.Match(bytes.ToLower(text))
	found := make([]string, 0, len(matches))
	seen := make(map[string]struct{}) // To avoid duplicate entries

	for _, match := range matches {
		keyword := match.MatchString()
		if _, exists := seen[keyword]; !exists {
			found = append(found, keyword)
			seen[keyword] = struct{}{}
		}
	}
	return found
}

// apkHandler handles apk archive formats.
type apkHandler struct {
	keywordMatcher *KeywordMatcher
	*defaultHandler
}

// newapkHandler creates an apkHandler.
func newAPKHandler() *apkHandler {
	return &apkHandler{
		defaultHandler: newDefaultHandler(apkHandlerType),
		keywordMatcher: getMatcher(),
	}
}
```

- Introduce a `KeywordMatcher` (name isn't super important) that’s initialized once during program startup. This matcher would load the keywords and exclusion list to build the trie, incurring the building cost just once.  
- Embed the matcher in the `apkHandler` for efficient lookups.  

With this setup, the `KeywordMatcher` could be seamlessly integrated into the `apkHandler` and streamline lookups.

It would then look like:
```
func (h *apkHandler) processDexClass(ctx logContext.Context, dexReader *dextk.Reader, node dextk.ClassNode, dexOutput *bytes.Buffer) {
    .......
	// Write the classOutput to the dexOutput
	dexOutput.Write(classOutput.Bytes())

	foundKeywords := h.keywordMatcher.FindKeywords(classOutput.Bytes()) <--- HERE

	// For each found keyword, create a keyword:value pair and append to dexOutput
	var keyValuePairs bytes.Buffer
	for str := range methodValues {
		for _, keyword := range foundKeywords {
			keyValuePairs.Reset()
			keyValuePairs.WriteString(keyword + ":" + str + "\n")
			dexOutput.Write(keyValuePairs.Bytes())
		}
	}
}
```

Reply:
Just implemented this in [454037b](https://github.com/trufflesecurity/trufflehog/pull/3517/commits/454037b983a9244a0f246b9587b34cc46755f688) and it's much faster. Really cool algorithm.

Reply:
That looks good to me. We can actually thank @zricethezav for introducing us to it.
[blog](https://trufflesecurity.com/blog/making-trufflehog-faster-with-aho-corasick) to go along.